Allow duosec to be used in nixos as a pam module.

sdier opened the pull request:
A bit of no idea dog here -- I'm not entirely sure I understand the testing requirements yet for a nixos-only change.  Please let me know if I misunderstood what to do.

Also, while I updated documentation in the nixos module, if additional documentation is needed please let me know where it should go.

Thanks!

###### Motivation for this change
Make it possible to configure duosec with pam for ssh.  It allows folks to use ssh keys without a 2FA prompt, but passwords with a 2FA prompt.

###### Things done

I tested this by installing settings on a machine with ssh with this configuration:
services.openssh.enable = true;
security.duosec.pam.enable = true;
security.duosec.autopush = true;
security.pam.services.sshd.duoSecurity.enable = true;

I did manual tests:
* Input of an invalid password does not prompt duo security.
* Input of a valid password creates a prompt to duo.
* Rejection of duo does not allow login.
* Repeated invalid passwords do not allow login.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

sdier committed:
nixos/security: Add duo-unix support to pam.
Also whitespace cleanup of surrounding code.

sdier committed:
nixos/security: Fix pam configuration file generation.

Infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -354,7 +354,7 @@ let
             || cfg.googleAuthenticator.enable
             || cfg.duoSecurity.enable)) ''
-              auth required pam_unix.so ${optionalString cfg.allowNullPassword "nullok"} likeauth
+              auth requisite pam_unix.so ${optionalString cfg.allowNullPassword "nullok"} likeauth
```
Not sure about this, you should read the comment above (line 344-349). Also pinging the original author @obadz and another person that seems to know their ways around PAM, @flokli

sdier replied:
Yeah, I forgot to change the comment, if folks end up agreeing I will edit the comment, but can wait until they get a chance to look.

I think requisite brings the right behavior when used in conjunction with additional challenge/response modules -- the behavior before this change isn't what I would have wanted. (if password was incorrect, required modules after this line were still being executed)

flokli replied:
I mostly did dig into the `account` module interface, less into the `auth` module interface.
So I'm not so sure about this either, and whether this might break other authentication modules (like pam_ldap).

Why is this so complicated in first place? Why do we call all pam modules two times? I wonder whether we might align with other distros do it here instead, they don't seem to load everything twice.

About avoiding to run duologin even when the password is wrong, pam writes the following for "required":

> If a 'required' module returns a status that is not 'success',
the operation will ultimately fail, but only after the modules below
it are invoked. This seems senseless at first glance I suppose, but
it serves the purpose of always acting the same way from the point
of view of the user trying to utilize the service. The net effect is
that it becomes impossible for a potential cracker to determine
which module caused the failure - and the less information a
malicious user has about your system, the better.

Which kinda makes sense too - so my suggestion would be to stick with keeping "auth required" for the scope of this PR, then as followup, cleanup the configuration duplication and add a early bailout switch for exiting before any 2FA, but after all primary authentication happened.

sdier replied:
This sounds good to me.  I will revert the part of the change with ‘auth required’ and think through a follow up change to pam.

For changes that span across packages in the nixos module is there a good place for me to post a proposal for comment before writing them — or is it preferred to have code first?

sdier replied:
Ok, this is ready for review again, thanks!

Infinisil commented on nixos/modules/security/pam.nix:
```diff
@@ -131,6 +131,18 @@ let
+          description = ''
+            If set, use the Duo Security pam module
+            <literal>pam_duo</literal> for authentication.  Requires
+            configuration of <literal>duo-unix</literal> package options.
```
What does "Requires configuration of `duo-unix` package options." mean? There's no such option in NixOS.

sdier replied:
I've changed it to say security.duosec options -- is this better?

Infinisil replied:
Oh I see, in that case it should be `<option>security.duosec</option>`

sdier replied:
Ok, thanks!

sdier committed:
nixos/security: Allow configuration of pam for duosec.

flokli reviewed (approved):
LGTM. Could you add an item inside `sec-release-19.03-new-services` in `nixos/doc/manual/release-notes/rl-1903.xml`?

sdier committed:
nixos/security: Add release note for duosec pam support for 19.03.

sdier commented:
@flokli I've added an item, PTAL -- I'm not sure if option hyperlinking supports linking in this form to support the '<name?>' syntax.  I also caught a silly typo in the description I had updated in 096e66a earlier today and fixed it.

Infinisil commented:
Looking good :)
